New API front and back to support RenderCommands

Mathys-Gasnier opened the pull request:
The current API doesn't make it easy to support RenderCommands in it's current state, making it mandatory to duplicate code/structs

So this new API should allow use on the two fronts. Some things were changed a bit, other were completly replaced.

PS. `Text` is still not working because we are waiting on what clay does with this

emoon commented:
I'm fine with duplicating structs when we have to. What I recommend we do is something like this

```Rust
#[repr(C)]
struct FFI {
    a: i32,
    b: f64,
    c: u8,
}

#[repr(C)]
struct RustStruct {
    a: i32,
    b: f64,
    c: u8,
}

fn offset_of<T, F>(field: &F, base: &T) -> usize {
    let base_ptr = base as *const T as *const u8;
    let field_ptr = field as *const F as *const u8;
    (field_ptr as usize) - (base_ptr as usize)
}

#[test]
fn verify_field_offsets() {
    let dummy_ffi = FFI {
        a: 0,
        b: 0.0,
        c: 0,
    };
    let dummy_rust = RustStruct {
        a: 0,
        b: 0.0,
        c: 0,
    };

    // Field offsets in FFI
    let ffi_a_offset = offset_of(&dummy_ffi.a, &dummy_ffi);
    let ffi_b_offset = offset_of(&dummy_ffi.b, &dummy_ffi);
    let ffi_c_offset = offset_of(&dummy_ffi.c, &dummy_ffi);

    // Field offsets in RustStruct
    let rust_a_offset = offset_of(&dummy_rust.a, &dummy_rust);
    let rust_b_offset = offset_of(&dummy_rust.b, &dummy_rust);
    let rust_c_offset = offset_of(&dummy_rust.c, &dummy_rust);

    // Assert that offsets match
    assert_eq!(ffi_a_offset, rust_a_offset);
    assert_eq!(ffi_b_offset, rust_b_offset);
    assert_eq!(ffi_c_offset, rust_c_offset);
}

fn main() {
    println!("Run `cargo test` to verify field offsets.");
}
```

This will verify that the structs always matches so when a new `clay.h` pop up we won't miss updating something if they have been moved around.

Mathys-Gasnier commented:
On my side i'm not ok with having two structs that do the same thing just in a different context, the value stored is the same... So why not having only one struct to store it.

Tests like those are not going to work, every fields are not a one to one representation because of how rust bindings work (See `BorderRadius` for example).
It might work on most struct, but some aren't going to like that, even more for struct that use enums...

emoon commented:
Right, I was thinking of the case if we need to duplicate them for some reason. I just don't want us to get into a case where the some structs doesn't match for some reason which can be very annoying to track down.

Mathys-Gasnier commented:
No i meant like for example you use `Clay_RectangleElementConfig` to create the rectangle, and it gets passed back on `Clay_RenderCommand.config`, so why have two struct to store a rectangle config instead of one, since they are pretty much the same thing

Mathys-Gasnier commented:
And my PR name isn't very good... I didn't change the API front, I switched around some things, but i changed a lot of the API "back-end" if we can call it that
